chain/client/src/test_utils/test_env_builder.rs

```diff
@@ -254,56 +255,61 @@ impl TestEnvBuilder {
-                        .block_producers_per_epoch(vec![ret.validators.clone()]);
-                    MockEpochManager::new_with_validators(
-                        ret.stores.as_ref().unwrap()[i].clone(),
-                        vs,
-                        ret.chain_genesis.epoch_length,
-                    )
-                    .into()
-                })
-                .collect();
-            assert!(
-                ret.shard_trackers.is_none(),
-                "Cannot override shard_trackers without overriding epoch_managers"
-            );
-            assert!(
-                ret.runtimes.is_none(),
-                "Cannot override runtimes without overriding epoch_managers"
-            );
-            ret.epoch_managers = Some(epoch_managers);
-            ret
+            return ret;
         }
+        let epoch_managers: Vec<EpochManagerKind> = (0..ret.clients.len())
+            .map(|i| {
+                let vs = ValidatorSchedule::new_with_shards(ret.num_shards.unwrap_or(1))
+                    .block_producers_per_epoch(vec![ret.validators.clone()]);
+                MockEpochManager::new_with_validators(
+                    ret.stores.as_ref().unwrap()[i].clone(),
+                    vs,
+                    ret.chain_genesis.epoch_length,
+                )
+                .into()
+            })
+            .collect();
+        assert!(
+            ret.shard_trackers.is_none(),
+            "Cannot override shard_trackers without overriding epoch_managers"
+        );
+        assert!(
+            ret.runtimes.is_none(),
+            "Cannot override runtimes without overriding epoch_managers"
+        );
+        ret.epoch_managers = Some(epoch_managers);
+        ret
     }
 
     /// Visible for extension methods in integration-tests.
-    pub fn internal_ensure_epoch_managers_for_nightshade_runtime(
+    pub fn internal_initialize_nightshade_runtimes(
         self,
-    ) -> (Self, Vec<PathBuf>, Vec<Store>, Vec<Arc<EpochManagerHandle>>, bool) {
-        let state_snapshot_enabled = self.state_snapshot_enabled;
-        let builder = self.ensure_epoch_managers();
-        let default_home_dirs =
-            (0..builder.clients.len()).map(|_| PathBuf::from("../../../..")).collect_vec();
-        let home_dirs = builder.home_dirs.clone().unwrap_or(default_home_dirs);
-        let stores = builder.stores.clone().unwrap();
-        let epoch_managers = builder
-            .epoch_managers
-            .clone()
-            .unwrap()
-            .into_iter()
-            .map(|kind| match kind {
+        runtime_configs: Vec<RuntimeConfigStore>,
+        nightshade_runtime_creator: impl Fn(
+            PathBuf,
+            Store,
+            Arc<EpochManagerHandle>,
+            RuntimeConfigStore,
+        ) -> Arc<dyn RuntimeAdapter>,
+    ) -> Self {
+        let builder = self.ensure_home_dirs().ensure_epoch_managers().ensure_stores();
+        let runtimes = multizip((
```

Comment:
Now we don't need to pass back 50 things to `nightshade_runtimes_with_runtime_config_store`